### PR TITLE
Generic SQL - connections not closed properly

### PR DESCRIPTION
**Pull request by rshunim**
fixes: https://jira-dc.paloaltonetworks.com/browse/XSUP-35920

### CODE REVIEW

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py [hunk 1]
```diff
@@ -598,6 +599,19 @@ def fetch_incidents(client: Client, params: dict):
 ]
 
 
+# Custom context manager to manage the connection with TTL
+@contextmanager
+def connection_with_ttl(engine, ttl_seconds):
```
Comment (dorschw) [inline Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:604]: add type hints

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py [hunk 2]
```diff
@@ -598,6 +599,19 @@ def fetch_incidents(client: Client, params: dict):
...
+    # Open a new connection
+    connection = engine.connect()
+    try:
+        # Yield the connection to the caller
+        yield connection
+    finally:
```
Comment (dorschw) [inline Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:610]: let's add an `except Exception` just to log whatever happened

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py [hunk 3]
```diff
@@ -629,41 +643,46 @@ def main():
...
-        if 'certificate verify failed' in str(err):
-            return_error("Unexpected error: certificate verify failed, unable to get local issuer certificate. "
-                         "Try selecting 'Trust any certificate' checkbox in the integration configuration.")
-        else:
-            return_error(
-                f'Unexpected error: {str(err)} \nquery: {demisto.args().get("query")}')
+
+        try:
+            # Use the custom context manager to manage the connection
+            with connection_with_ttl(engine.connection, ttl_seconds=engine.pool_ttl) as connection:
```
Comment (dorschw) [inline Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:653]: I assume we have a `pool_ttl` even if not using a pool?

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py [hunk 4]
```diff
@@ -629,41 +643,46 @@ def main():
...
+                commands: dict[str, Callable[[Client, dict[str, str], str], tuple[str, dict[Any, Any], list[Any]]]] = {
+                    'test-module': test_module,
+                    'query': sql_query_execute,
+                    'pgsql-query': sql_query_execute,
+                    'sql-command': sql_query_execute
+                }
```
Comment (dorschw) [inline Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:659]: Let's pull this out of the `try` block, keeping it minimal

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py [hunk 5]
```diff
@@ -629,41 +643,46 @@ def main():
...
+                return_error("Unexpected error: certificate verify failed, unable to get local issuer certificate. "
+                             "Try selecting 'Trust any certificate' checkbox in the integration configuration.")
+            return_error(f'Unexpected error: {str(err)} \nquery: {demisto.args().get("query")}')
     finally:
         try:
-            if client.connection:
-                client.connection.close()
+            if engine.connection:
+                # Close the engine after use
+                engine.connection.dispose()
```
Comment (dorschw) [inline Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:678]: what's the difference between `close` and `dispose`?

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py [hunk 6]
```diff
@@ -629,41 +643,46 @@ def main():
...
             demisto.error(f'Failed closing connection: {str(ex)}')
-        if sql_loggers:
-            for lgr in sql_loggers:
-                demisto.debug(f'setting back ERROR for logger: {repr(lgr)}')
-                lgr.setLevel(logging.ERROR)
+
+            if sql_loggers:
+                for lgr in sql_loggers:
+                    demisto.debug(f'setting back ERROR for logger: {repr(lgr)}')
+                    lgr.setLevel(logging.ERROR)
```
Comment (dorschw) [inline Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:685]: why indent?

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.yml [hunk 7]
```diff
@@ -104,8 +104,8 @@ configuration:
   type: 13
   required: false
 description: 'Use the Generic SQL integration to run SQL queries on the following databases: MySQL, PostgreSQL, Microsoft SQL Server, and Oracle.'
-display: Generic SQL
-name: Generic SQL
+display: Generic SQL - dev
+name: Generic SQL - dev
```
Comment (dorschw) [inline Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.yml:108]: make sure to remove before merging :)

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py (hunk 3 above)
Comment (rshunim) [reply Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:653]: Yes, you're.
The default is defined to be 600 seconds

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py (hunk 5 above)
Comment (rshunim) [reply Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:678]: image

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py (hunk 6 above)
Comment (rshunim) [reply Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:685]: It's the same as in the origin master, I don't know why it seems like a change

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.yml (hunk 7 above)
Comment (rshunim) [reply Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.yml:108]: sure

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py [hunk 8]
```diff
@@ -1,5 +1,5 @@
-import demistomock as demisto
-from CommonServerPython import *
+import demistomock as demisto  # noqa: F401
+from CommonServerPython import *  # noqa: F401
```
Comment (dorschw) [inline Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:2]: is this change required? It worked before without these.

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py [hunk 9]
```diff
@@ -31,6 +31,7 @@
 pymysql.install_as_MySQLdb()
 
 GLOBAL_CACHE_ATTR = '_generic_sql_engine_cache'
+ENGINES = '_generic_sql_engines'
```
Comment (dorschw) [inline Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:34]:
```suggestion
GLOBAL_ENGINE_CACHE_ATTR = '_generic_sql_engines'
```

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py [hunk 10]
```diff
@@ -165,8 +167,16 @@ def _create_engine_and_connect(self) -> sqlalchemy.engine.base.Connection:
             cache_key = self._get_cache_string(str(db_url), ssl_connection)
             engine = cache.get(cache_key, None)
             if engine is None:  # (first time or expired) need to initialize
+
+                engines = getattr(sqlalchemy, ENGINES, {})
```
Comment (dorschw) [inline Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:171]:
```suggestion
                cached_engines = getattr(sqlalchemy, ENGINES, {})
```

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py [hunk 11]
```diff
@@ -165,8 +167,16 @@ def _create_engine_and_connect(self) -> sqlalchemy.engine.base.Connection:
...
+
+                if cache_key in engines:
+                    engines[cache_key].dispose()
+                    engines.pop(cache_key)
+
                 engine = sqlalchemy.create_engine(db_url, connect_args=ssl_connection)
                 cache[cache_key] = engine
+                engines[cache_key] = engine
```
Comment (dorschw) [inline Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:179]:
```suggestion
                engines[cache_key] = engine  # Keep in cache to allow disposing later
```

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py [hunk 12]
```diff
@@ -165,8 +167,16 @@ def _create_engine_and_connect(self) -> sqlalchemy.engine.base.Connection:
             cache_key = self._get_cache_string(str(db_url), ssl_connection)
             engine = cache.get(cache_key, None)
             if engine is None:  # (first time or expired) need to initialize
+
+                engines = getattr(sqlalchemy, ENGINES, {})
+
+                if cache_key in engines:
+                    engines[cache_key].dispose()
```
Comment (dorschw) [inline Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:174]: How does SQLA behave if we call `.dispose` _after_ the connection is closed (say, it's been a day since the last connection)?

Comment (dorschw) [inline Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:174]: do we want to dispose the engine on each command, or only if it's beyond its ttl?

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py
Comment (dorschw) [inline Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:1]: Please add a simple unit test for the new mechanism.

**Review by dorschw — COMMENTED** [review]
Nice job

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py (hunk 8 above)
Comment (rshunim) [reply Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:2]: Actually I didn't add it, but I see the master does not contain those lines, so I'll revert it

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py (hunk 9 above)
Comment (rshunim) [reply Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:34]: Done

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py (hunk 10 above)
Comment (rshunim) [reply Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:171]: Done

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py (hunk 12 above)
Comment (rshunim) [reply Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:174]: It's fine, since the .dispose() method is an engine's method, you are actually disposing of the engine's object, which contains the connections.

Comment (rshunim) [reply Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:174]: only if it's beyond its ttl

Comment (dorschw) [reply Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py]: where in the code?

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py
Comment (rshunim) [reply Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:1]: Two UT's were added

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py [hunk 13]
```diff
@@ -165,8 +167,16 @@ def _create_engine_and_connect(self) -> sqlalchemy.engine.base.Connection:
             cache_key = self._get_cache_string(str(db_url), ssl_connection)
             engine = cache.get(cache_key, None)
             if engine is None:  # (first time or expired) need to initialize
+
+                cached_engines = getattr(sqlalchemy, GLOBAL_ENGINE_CACHE_ATTR, {})
+
+                if cache_key in cached_engines:
```
Comment (dorschw) [inline Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:173]:
```suggestion
                if cache_key in cached_engines:
                # engine is None, but cache_key in cached_engines, so the ttl must have passed - let's dispose the engine to make sure the connection is closed.
```

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL_test.py [hunk 14]
```diff
@@ -626,3 +639,58 @@ def test_generate_variable_names_and_mapping(bind_variables_values_list: list, q
...
+
+
+def test_create_engine_and_connect_new_engine(mocker):
+    """
+    Given
+    - No engine is in the cache
+    When
+    - running create_engine_and_connect
+    Then
+    - Ensure a new engine is created
```
Comment (dorschw) [inline Packs/GenericSQL/Integrations/GenericSQL/GenericSQL_test.py:679]:
```suggestion
    - Ensure a new engine is created and replaces the old one in cache
```

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL_test.py [hunk 15]
```diff
@@ -626,3 +639,58 @@ def test_generate_variable_names_and_mapping(bind_variables_values_list: list, q
...
+    - Ensure the old engine is disposed and removed from the cache
```
Comment (dorschw) [inline Packs/GenericSQL/Integrations/GenericSQL/GenericSQL_test.py:680]: make sure `dispose` is really called

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL_test.py [hunk 16]
```diff
@@ -626,3 +639,58 @@ def test_generate_variable_names_and_mapping(bind_variables_values_list: list, q
...
+    setattr(sqlalchemy, GLOBAL_ENGINE_CACHE_ATTR, {'1': old_engine})
+    mock_create_engine = mocker.patch.object(sqlalchemy, 'create_engine', return_value=Engine('New Engine'))
+
+    client = Client(dialect='Teradata', host='host', username='username', password='password', port='', connect_parameters='',
+                    database='', ssl_connect=False, use_pool=True)
+
+    mock_create_engine.assert_called_once()
+    cache = getattr(sqlalchemy, GLOBAL_ENGINE_CACHE_ATTR)
+    assert client.connection == 'New Engine'
+    assert cache['1'] != old_engine
```
[the remark this answers is not in the document]
Comment (rshunim) [reply Packs/GenericSQL/Integrations/GenericSQL/GenericSQL_test.py:701]: added another assert as we talk about

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL_test.py (hunk 15 above)
Comment (rshunim) [reply Packs/GenericSQL/Integrations/GenericSQL/GenericSQL_test.py:684]: added another check for this

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL_test.py [hunk 17]
```diff
@@ -626,3 +639,58 @@ def test_generate_variable_names_and_mapping(bind_variables_values_list: list, q
     result = generate_variable_names_and_mapping(bind_variables_values_list, query, dialect)
     assert expected_result[0] == result[0]
     assert expected_result[1] == result[1]
+
+
+def create_engine_and_connect_init(mocker, global_cache, cache_string):
```
[the remark this answers is not in the document]
Comment (rshunim) [reply Packs/GenericSQL/Integrations/GenericSQL/GenericSQL_test.py]: added it in the new class

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL_test.py [hunk 18]
```diff
@@ -10,6 +10,32 @@
...
+    """
+
+    def __init__(self, mocker, global_cache: dict, cache_string: str, dialect: str, host: str, username: str, password: str,
+                 port: str, database: str, connect_parameters: str, ssl_connect: bool, use_pool: bool):
+
+        mocker.patch.object(Client, '_convert_dialect_to_module')
+        mocker.patch.object(Client, '_generate_db_url')
+        mocker.patch.object(Client, '_get_global_cache', return_value=global_cache)
+        mocker.patch.object(Client, '_get_cache_string', return_value=cache_string)
+        super().__init__(dialect, host, username, password, port, database, connect_parameters, ssl_connect, use_pool)
```
Comment (dorschw) [inline Packs/GenericSQL/Integrations/GenericSQL/GenericSQL_test.py:25]:
```suggestion
        def _convert_dialect_to_module:
            ...            
```

and the same for all others, overriding them

Comment (rshunim) [reply Packs/GenericSQL/Integrations/GenericSQL/GenericSQL_test.py]: Done

File: Packs/GenericSQL/pack_metadata.json [hunk 19]
```diff
@@ -3,7 +3,7 @@
     "description": "Connect and execute sql queries in 4 Databases: MySQL, PostgreSQL, Microsoft SQL Server and Oracle",
     "support": "xsoar",
     "serverMinVersion": "5.0.0",
-    "currentVersion": "1.2.4",
+    "currentVersion": "1.2.5",
```
Comment (dorschw) [inline Packs/GenericSQL/pack_metadata.json:6]: worth a minor bump, please rename `1_2_5.md` to `1_3_0.md` accordingly

```suggestion
    "currentVersion": "1.3.0",
```

**Review by dorschw — APPROVED** [review]
